packages/zoe/src/contracts/attestation/expiringNFT.js

```diff
@@ -0,0 +1,264 @@
+      return empty;
+    }
+
+    const lienedAtStart = lienedAmounts.get(address);
+    const notExpired = lienedAtStart.filter(({ amountLiened, expiration }) => {
+      if (hasExpired(expiration, currentTime)) {
+        return false;
+      }
+      totalStillLiened = AmountMath.add(totalStillLiened, amountLiened);
+      return true;
+    });
+    // commit point
+
+    // Remove from cannotGetExtension since there is no lien left
+    if (AmountMath.isEmpty(totalStillLiened)) {
+      cannotGetExtension.delete(address);
+    }
+
+    lienedAmounts.set(address, notExpired);
+    return totalStillLiened;
+  };
+
+  /** @type {getLienAmount} */
+  const getLienAmount = async (address, currentTime) => {
+    assert.typeof(address, 'string');
+
+    // We should first unlien any amounts for which the lien has
+    // expired, then return the total lien still remaining for the address
+    return unlienIfExpired(address, currentTime);
+  };
+
+  /** @type {ExtendExpiration} */
+  const extendExpiration = (seat, newExpiration) => {
+    const oldAttestationAmount = seat.getAmountAllocated('Attestation');
+
+    const attestationValue =
+      /** @type {SetValue} */ (oldAttestationAmount.value);
+
+    // TODO: allow for multiple elements in the amount value to each
+    // be extended. Currently, we restrict the value to a single element.
+    assert(
+      attestationValue.length === 1,
+      X`We can currently only extend a single attestation element at a time`,
+    );
+    const {
+      expiration,
+      handle,
+      address,
+      amountLiened,
+    } = /** @type {ExpiringAttElem} */ (attestationValue[0]);
+
+    assert(
+      !cannotGetExtension.has(address),
+      `The address ${address} cannot extend the expiration for attestations`,
+    );
+
+    assert(
+      newExpiration > expiration,
+      `The newExpiration ${newExpiration} must be later than the old expiration ${expiration}`,
+    );
+    // Importantly, we cannot drop the uniqueHandle and make a new
+    // one, because that would allow someone to escrow the newly
+    // minted attestation in a contract and the contract would have no
+    // way of knowing it was representing the same value as before,
+    // allowing the user to misrepresent the liened amount as larger
+    // then it actually is.
+    const newAttestationElem = makeAttestationElem(
+      address,
+      amountLiened,
+      newExpiration,
+      handle,
+    );
+    const amountToMint = AmountMath.make(attestationBrand, [
+      newAttestationElem,
+    ]);
+
+    // commit point within updateLien
+    updateLien(newAttestationElem);
+    zcfMint.burnLosses(seat.getCurrentAllocation(), seat);
+    zcfMint.mintGains({ Attestation: amountToMint }, seat);
+    seat.exit();
+  };
+
+  // This is only released when the lien is empty
+  /** @type {DisallowExtensions} */
+  const disallowExtensions = address => {
+    assert.typeof(address, 'string');
+    cannotGetExtension.add(address);
```

Comment:
After slashing, it looks like you cannot extend current liens - you can only ask for new liens. But you also cannot extend those new liens unless you allow all your liens to expire. The slash event might be inconsequential to the existing liens. Would it be possible just to check the account state when an extension is requested to make sure the extension meets the criteria one would have for a new lien? If so, then you could ignore slashing entirely. Of course, I might not understand what extension is about, or the constraints on it.

Reply:
Huh, very interesting idea. 

The extensions are doing something special: making a new attestation payment with the same unique identifier as the old attestation payment. This is the identifier that voting contracts will use to know that the same fungible amount isn't being double counted. 

I suspect that allowing an extension just if the criteria for a new loan are met is problematic. Let's say that Alice has an attestation for 50 BLD that expires tomorrow, and an attestation for 100 BLD that expires next week. Slashing occurs such that Alice only has 125 BLD bonded (I made up this number - this is not applying any slashing calculation). Now Alice tries to extend both of her expirations. The first extension will succeed, and the second will fail. But Alice probably didn't anticipate that, and didn't think about the ordering of her extensions. 

I think it's cleaner to just turn off extensions in the case of slashing altogether, so the user doesn't need to think about fine-grained details like this. 

> But you also cannot extend those new liens unless you allow all your liens to expire.

Actually, you can create as many new liens as you want, as long as the prerequisites are met. So as soon as a single lien expires (and a RUN LoC attestation of at least that amount has been redeemed), you can create a new lien.